test/controllers/tradingNames/AddTradingNameControllerSpec.scala

```diff
@@ -0,0 +1,177 @@
+
+        val view = application.injector.instanceOf[AddTradingNameView]
+
+        val list = TradingNameSummary.addToListRows(baseAnswers, waypoints, AddTradingNamePage())
+
+        status(result) mustBe OK
+        contentAsString(result) mustBe view(form, waypoints, list, canAddTradingNames = true)(request, messages(application)).toString
+      }
+    }
+
+    "must return OK and the correct view for a GET when the maximum number of trading names have already been added" in {
+
+      val userAnswers =
+        basicUserAnswersWithVatInfo
+          .set(TradingNamePage(Index(0)), TradingName("foo")).success.value
+          .set(TradingNamePage(Index(1)), TradingName("foo")).success.value
+          .set(TradingNamePage(Index(2)), TradingName("foo")).success.value
+          .set(TradingNamePage(Index(3)), TradingName("foo")).success.value
+          .set(TradingNamePage(Index(4)), TradingName("foo")).success.value
+          .set(TradingNamePage(Index(5)), TradingName("foo")).success.value
+          .set(TradingNamePage(Index(6)), TradingName("foo")).success.value
+          .set(TradingNamePage(Index(7)), TradingName("foo")).success.value
+          .set(TradingNamePage(Index(8)), TradingName("foo")).success.value
+          .set(TradingNamePage(Index(9)), TradingName("foo")).success.value
+
+      val application = applicationBuilder(userAnswers = Some(userAnswers)).build()
+
+      running(application) {
+        val request = FakeRequest(GET, addTradingNameRoute)
+
+        val view = application.injector.instanceOf[AddTradingNameView]
+
+        val list = TradingNameSummary.addToListRows(userAnswers, waypoints, AddTradingNamePage())
+
+        val result = route(application, request).value
+
+        status(result) mustBe OK
+        contentAsString(result) mustBe view(form.fill(true), waypoints, list, canAddTradingNames = false)(request, messages(application)).toString
+      }
+    }
+
+    "must save the answer and redirect to the next page when valid data is submitted" in {
+
+      val mockSessionRepository = mock[AuthenticatedUserAnswersRepository]
+
+      when(mockSessionRepository.set(any())) thenReturn true.toFuture
+
+      val application =
+        applicationBuilder(userAnswers = Some(baseAnswers))
+          .overrides(
+            bind[AuthenticatedUserAnswersRepository].toInstance(mockSessionRepository)
+          )
+          .build()
+
+      running(application) {
+        val request =
+          FakeRequest(POST, addTradingNameRoute)
+            .withFormUrlEncodedBody(("value", "true"))
+
+        val result = route(application, request).value
+        val expectedAnswers = baseAnswers.set(AddTradingNamePage(Some(index)), true).success.value
+
+        status(result) mustBe SEE_OTHER
+        redirectLocation(result).value mustBe AddTradingNamePage(Some(index)).navigate(waypoints, baseAnswers, expectedAnswers).url
+        verify(mockSessionRepository, times(1)).set(eqTo(expectedAnswers))
```

Comment:
I like the loose expectation/tight verification approach with mockito as it returns nulls on mismatches which causes the worst type of errors. Scalamock fails nicer as it fails detailing the expectation in the test. 

Sometimes I just find loose Mockito expectations with no tight verification. That makes me sad as it leads to gaps in what is tested :)

Reply:
Not to sure what you mean here but happy to discuss.

Reply:
No need for discussion :)

I have seen the 

```
when(mockSessionRepository.set(any())) thenReturn true.toFuture
```

pattern used a lot, checking for any()) in the when so anything passes which helps with not getting null pointer exceptions when the expectancy does not match, but then there is no verify to actually test the value that should have been passed. 

I can go in the code and just pass null everywhere and the test still passes etc. Having the tight verify with the actual expected value stops this.

```scala
verify(mockSessionRepository, times(1)).set(eqTo(expectedAnswers))
```

Just saying what is done is good practice in Mockito even if it is a bit more work to do than 

```
when(mockSessionRepository.set(eqTo(expectedAnswers)) thenReturn true.toFuture
```
which will null pointer in code if expectedAnswers is anything different to expected.

Reply:
Ah yes, I understand, I just wasn't sure initially what you were referring to but makes sense now. :-)